.azure-pipelines/testscripts_analyse/analyse_testcases.py

```diff
@@ -0,0 +1,141 @@
+from datetime import datetime
+from constant import DATAPLANE_FEATURES, PR_TOPOLOGY_TYPE
+
+
+def collect_all_scripts():
+    '''
+    This function collects all test scripts under the folder 'tests/'
+    and get the topology type marked in the script
+
+    The return value is a dict contains the script name and topology type
+    [{
+        'testscript': 'acl/custom_acl_table/test_custom_acl_table.py',
+        'topology': 't0'
+    }]
+    '''
+    location = sys.argv[1]
+
+    # Recursively find all files starting with "test_" and ending with ".py"
+    # Note: The full path and name of files are stored in a list named "files"
+    files = []
+    for root, dirs, file in os.walk(location):
+        for f in file:
+            if f.startswith("test_") and f.endswith(".py"):
+                files.append(os.path.join(root, f))
+    files = natsorted(files)
+
+    # Open each file and search for regex pattern
+    pattern = re.compile(r"[^@]pytest\.mark\.topology\(([^\)]*)\)")
+    test_scripts = []
+
+    for f in files:
+        # Remove prefix from file name:
+        filename = f[len(location) + 1:]
+        try:
+            with open(f, 'r') as file:
+                for line in file:
+                    # Get topology type of script from mark `pytest.mark.topology`
+                    match = pattern.search(line)
+                    if match:
+                        for topology in match.group(1).split(","):
+                            result = {
+                                "testscript": filename,
+                                "topology": topology.strip().strip('"').strip('\'')
+                            }
+                            test_scripts.append(result)
+        except Exception as e:
+            logging.error('Failed to load file {}, error {}'.format(f, e))
+
+    return test_scripts
+
+
+def check_PRChecker_coverd(test_scripts):
+    '''
+    Check if a script is included in the PR checker for the corresponding topology type
+    '''
+    pr_test_scripts_file = os.path.join(os.path.dirname(os.path.realpath(__file__)), "../", "pr_test_scripts.yaml")
+
+    # Get all the scripts included in different PR checker
+    try:
+        with open(pr_test_scripts_file) as f:
+            pr_test_scripts = yaml.safe_load(f)
+    except Exception as e:
+        logging.error('Failed to load file {}, error {}'.format(f, e))
+
+    # Check if a script is included in the PR checker for the corresponding topology type
+    # If the topology mark is "any", we will check if it is included in all topology types of PR checker.
+    i = 0
+    while i < len(test_scripts):
+        topo_type = test_scripts[i]["topology"]
+
+        if topo_type == "any":
+            for topology in PR_TOPOLOGY_TYPE:
+                test_scripts.append({
+                    "testscript": test_scripts[i]["testscript"],
+                    "topology": topology
+                })
+            test_scripts.remove(test_scripts[i])
```

Comment:
Use a separate list to store the filtered/expanded results instead of using in place list operation.
In place list operation is hard to understand and prune to mistakes.

Reply:
I haved created a new list `expanded_test_scripts`